### ISSUE TITLE
Confusing method matching for constructors

### ISSUE THREAD

**Issue by timholy** [JuliaLang/julia#9970]
This is getting greedy, but given that we have this lovely new "closest candidates" method-matching, I'm wondering if the match in situations like these can be improved:
```julia
julia> type ArrayContainer{T,N,I<:(Union(Int,UnitRange{Int})...)} <: AbstractArray{T,N}
           data::Array{T,N}
           indexes::I
       end

julia> ArrayContainer(zeros(3,3), (:,:))
ERROR: MethodError: `convert` has no method matching convert(::Type{ArrayContainer{T,N,I<:(Union(UnitRange{Int64},Int64)...,)}}, ::Array{Float64,2}, ::(Colon,Colon))
This may have arisen from a call to the constructor ArrayContainer{T,N,I<:(Union(UnitRange{Int64},Int64)...,)}(...), since type constructors fall back to convert methods in julia v0.4.
Closest candidates are:
  convert(::Type{Ptr{Void}}, ::Array{T,N})
  convert{T}(::Type{Ptr{T}}, ::Array{T,N})
  convert{T,N}(::Type{AbstractArray{T,N}}, ::AbstractArray{T,N})
  ...

 in call at base.jl:35
```
Probably the most straightforward problem is that the conflicting messages are a bit confusing, and for whatever reason my eye is drawn to the "Closest candidates" list more readily than the "This may have arisen..." line. (That's probably just me, however.) One option would be to suppress the candidates list in that case; perhaps another be to emphasize the match to the first argument in such cases.

**Comment by dhoegh** [comment]
It should not be a problem to special case `convert` and `call` to only show if the first argument is matched.

**Comment by dhoegh** [comment]
I really wonder what the best solution to this is. The special casing `convert` and `call` seems weird, what with `getindex` and so on. Maybe the best solution is to introduce a weighting on the first two/three input arguments to weight the methods with a match on the first argument higher than a match on the latter arguments.

**Comment by ivarne** [comment]
A general solution is always prefered, but this is not APIs that users will program too. I don't think special casing a dozen functions with special syntax is weird, especially if it makes Julia more pleasant for users.

**Comment by dhoegh** [comment]
We could also save the match pattern for the top three methods which will be an `Array{Bool,1}` where they could be sorted on number of matches and order of arguments.

**Comment by timholy** [comment]
I'm not sure what's best either, but I'm not against special-casing. Treating arguments differently by their position in general smacks of single-dispatch rather than multiple-dispatch.

**Comment by simonster** [comment]
It seems like we should pool `call` and `convert` methods when showing results for MethodErrors for `convert`. `call` never throws a MethodError itself due to the `convert` fallback, so otherwise I don't think you'll ever get a meaningful method list for a constructor. In this case I think the constructor would be at the top of the list if we made that change, since two arguments match.

**Comment by dhoegh** [comment]
I have a fix for this issue in #10129
